Step2 로또 자동 구현

gmltkddl1 opened the pull request.

gmltkddl1 committed:
refactor: step1 피드백 반영

gmltkddl1 committed:
feat: step2 로또(자동) 구현

요구사항 '로또 구매금액을 입력받는다' 구현

gmltkddl1 committed:
feat: step2 로또(자동) 구현

요구사항 '로또 한장의 가격은 1000원이다' 구현

gmltkddl1 committed:
feat: step2 로또(자동) 구현

요구사항 '로또 구매 갯수는 구매금액을 한장 가격으로 나눈 값이다' 구현

gmltkddl1 committed:
feat: step2 로또(자동) 구현

요구사항 '로또 번호는 1부터 45까지 번호 중 6개의 랜덤 조합이다' 구현

gmltkddl1 committed:
feat: step2 로또(자동) 구현

요구사항 '로또 구매 갯수만큼 로또 번호를 생성한다' 구현

gmltkddl1 committed:
feat: step2 로또(자동) 구현

요구사항 '당첨 번호를 입력받는다' 구현

gmltkddl1 committed:
feat: step2 로또(자동) 구현

요구사항 '당첨 번호과 로또 번호들을 비교하여 번호의 일치 갯수를 구한다' 구현

gmltkddl1 committed:
feat: step2 로또(자동) 구현

요구사항 '갯수별 금액과 갯수를 곱해 수익금을 만든다' 구현

gmltkddl1 committed:
feat: step2 로또(자동) 구현

요구사항 '갯수별 금액과 갯수를 곱해 수익금을 만든다' 구현

gmltkddl1 committed:
feat: step2 로또(자동) 구현

요구사항 '갯수별 금액과 갯수를 곱해 수익금을 만든다' 구현

gmltkddl1 committed:
feat: step2 로또(자동) 구현

요구사항 '갯수별 금액과 갯수를 곱해 수익금을 만든다' 구현

gmltkddl1 committed:
feat: step2 로또(자동) 구현

로또 출력 구현 및 리팩토링

wisemuji commented on src/main/kotlin/lotto/domain/RandomSetGenerator.kt:
```diff
@@ -0,0 +1,11 @@
+    override fun getSet(): List<Int> {
+        return (MINIMUM_NUMBER..MAXIMUM_NUMBER).shuffled().subList(0, NUMBER_OF_NUMBER)
+    }
+}
```
일종의 전략 패턴을 활용하여 테스트 가능하게 구현해주셨네요 👍 
테스트가 어려운 랜덤 로직을 참조하지 않게 되었어요.

이미 잘 알고 계시는 것 같지만, 다음 글을 읽어보시면서 전략 패턴의 효용을 정리해보셔도 좋을 것 같아요.
- https://tecoble.techcourse.co.kr/post/2021-10-04-strategy-command-pattern/
- https://tecoble.techcourse.co.kr/post/2020-05-07-appropriate_method_for_test_by_parameter/

wisemuji replied:
또한 한 가지 고민해볼 것이 있습니다.

엘레강트 오브젝트라는 책의 **`-er`로 끝나는 이름을 사용하지 마세요** 파트에서는 다음과 같이 설명합니다.

> 클래스의 이름은 무엇을 하는지에 대한게 아닌 무엇인지에 기반을 두고 지어야 한다
올바른 이름을 짓기 위해선 클래스의 객체들이 무엇을 캡슐화할 것인지 관찰 후 적합한 이름을 찾아야 한다
외부에서 객체에게 무엇을 하도록 메세지를 던지면 객체 스스로 무엇을 할 지를 정해야 한다
여기서 -er로 끝나는 이름을 짓는다면, 그것은 바로 기능에 기반하여 클래스를 명명했다고 볼 수 있다
[Bad]
Manager, Controller, Handler, Converter
[Good]
Target, Content, EncodedText

솔직히 말씀드리면 저는 이 내용에 대해 온전히 동의하진 않아요! (ㅋㅋ)

그럼에도 언급하는 이유는, `-er`로 끝나지 않는 것이 중요하기보단,
무엇을 하는지에 초점을 맞추지 말고, 무엇인지 그 자체를 객체로 만들어보라는 핵심이 중요하기 때문입니다.

wisemuji replied:
객체지향이란 현실 세계의 모방, 추상화 라는 말이 있습니다. 즉, 현실 세계를 모방해서 단순화(추상화)한다는 의미가 숨어있습니다.

Lotto 객체를 생성하는 책임을 갖는 객체를 어떻게 만들어볼 수 있을까요?
지금처럼 `Generator`와 같은 객체를 만들어 Lotto 리스트를 반환하게 만들 수 있겠네요.
하지만 이러한 객체는 “자신의 상태를 스스로 관리하는 자율적인 존재“인 능동적인 객체라고 보기는 어렵다고 생각합니다. 무엇을 하는지에 강하게 묶여 있어 추가 책임을 가질 수도, 스스로 판단하여 메세지를 주고받을 수 있는 자유도 없습니다.

저라면 현실세계를 모방한 `LottoMachine`과 같은 객체를 만들어 Lotto 객체를 생성하는 책임을 부여할 것 같아요.

물론 실제로 현업을 진행하다보면 -er로 끝나는 이름을 짓는 경우(즉, 기능에 기반하여 클래스를 명명하는 경우)가 어쩔 수 없이 발생하기도 합니다.
지금은 객체지향적 설계에 대한 학습이 주가 되는 미션인만큼, 이번 기회에 이러한 메세지를 던지는 구조와 설계를 학습하고 적용해보면 어떨까 하여 드린 피드백입니다 :slightly_smiling_face:

wisemuji commented on README.md:
기능 목록을 잘 도출해주시고, 이 기능 목록이 커밋 단위로도 이어졌네요 👏

wisemuji commented on src/main/kotlin/lotto/domain/Lotto.kt:
```diff
@@ -0,0 +1,15 @@
+
+    fun isLotto(): Boolean {
+        return numbers.all { it in (MINIMUM_NUMBER..MAXIMUM_NUMBER) }
+    }
```
테스트 코드에서만 활용되는 함수가 실제 프로덕션 코드에 포함되어 있는 것이 적절할까요?

wisemuji commented on src/test/kotlin/lotto/LottoGeneratorTest.kt:
```diff
@@ -0,0 +1,15 @@
+class LottoGeneratorTest {
+    @Test
+    fun `로또 구매 갯수만큼 로또 번호를 생성한다`() {
+        val lottoSeller = LottoSeller(5000)
```
저는 금액을 노출할 때에는 가독성을 위해 다음과 같이 표현합니다. 참고만 해주셔도 좋습니다.
```suggestion
        val lottoSeller = LottoSeller(5_000)
```

wisemuji commented on src/test/kotlin/lotto/LottoNumberMatcherTest.kt:
```diff
@@ -0,0 +1,25 @@
+                override fun getSet(): List<Int> {
+                    return listOf(1, 2, 3, 4, 7)
+                }
+            }
```
이번 기회에 [functional interface](https://kotlinlang.org/docs/fun-interfaces.html)를 적용해보시면 어떨까요?

wisemuji commented on src/main/kotlin/lotto/domain/Lotto.kt:
```diff
@@ -0,0 +1,15 @@
+package lotto.domain
+
+class Lotto(setGenerator: SetGenerator = RandomSetGenerator()) {
```
`SetGenerator`, `RandomSetGenerator`라는 이름을 보면 제너럴한 Set을 리턴할 것으로 기대할 수 있는데요, 실제로는 로또와 종속되어 있는 객체입니다.

다른 개발자들도 작동 방식을 이해할 수 있도록 수정해보면 어떨까요?

wisemuji commented on src/test/kotlin/lotto/LottoTest.kt:
```diff
@@ -0,0 +1,12 @@
+        val lotto = Lotto()
+        lotto.numbers.all { it in (1..45) }
+    }
+}
```
랜덤 로직을 참조하고 있는 테스트는 매번 실행할때마다 어떤 결과가 나올지 보장할 수 없습니다.

특정 숫자가 생성되었을 때에 버그가 일어난다고 가정한다면, 테스트를 통해 이러한 버그를 발견하지 못할 확률이 있다는 것, 즉 간헐적으로 실패할 수 있다는 것입니다.

wisemuji commented on src/main/kotlin/lotto/domain/LottoSeller.kt:
LottoSeller와 LottoResult를 현실세계에 존재한다고 생각해볼까요?
```
main이 LottoSeller에게 LottoResult를 요청한다
```
이 로직은
```
사용자가 로또 판매자에게 로또 결과를 요청한다
```
이렇게 해석해볼 수 있을 것 같아요. 조금 어색하게 느껴지시죠?

1. 로또 판매자가 로또 결과 및 수익률을 판단하는 것이 적절할까요?
2. 사용자가 로또 판매자에게 로또 결과를 요청하는 것이 적절할까요?

물론 객체지향적으로 설계할 때 모든 객체를 완벽하게 현실세계에 빗댈 수는 없습니다. 하지만 객체를 스스로 생각하고 스스로 결정할 수 있도록 현실 세계의 생명체에 비유하는 것은 상태와 행위를 캡슐화하는 소프트웨어 객체의 자율성을 설명하는 데 효과적입니다.

이번 단계에서는 이 코멘트 반영에 초점을 맞춰보시는 것을 추천드립니다.

wisemuji reviewed (changes requested):
안녕하세요 희상님!

로또 자동 구현을 잘 해주셨습니다 💯  
객체 설계 측면에서 몇 가지 고민해보시면 좋을만한 코멘트를 달았으니 확인해주시면 감사하겠습니다.

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏 💪

gmltkddl1 committed:
refactor: step2 피드백 반영

동작 위주의 클래스에서 사물위주의 클래스 설계로 변경
테스트용 코드 삭제
functional interface로 수정
SetGenerator -> LottoNumberGenerator로 변경
랜덤 테스트 삭제
LottoSeller에서 일부 책임을 Bank로 분리

wisemuji commented on src/main/kotlin/lotto/domain/Bank.kt:
```diff
@@ -0,0 +1,32 @@
+package lotto.domain
+
+class Bank {
+    fun getLottoResult(
+        lottos: List<Lotto>,
+        winningNumbers: WinningNumbers,
+        purchasePrice: Int,
+    ): LottoResult {
+        val resultMap = LottoResult.getResultMap(lottos, winningNumbers)
+        val profitMap = getProfitMap(resultMap)
+        val profit = getProfit(profitMap)
+        val profitRate = getProfitRate(profit, purchasePrice)
+
+        return LottoResult(lottos, resultMap, profitRate)
+    }
```
`Bank`에게 `LottoResult`를 요청하는 것이 아닌, `LottoResult` 스스로 객체를 생성하는 책임을 가질 수 있도록 개선해보면 어떨까요? 지금 구조에서 크게 변경하지 않아도 될 것 같아요.

wisemuji commented on src/main/kotlin/lotto/domain/LottoResult.kt:
```diff
@@ -0,0 +1,27 @@
+        fun getResultMap(
+            lottos: List<Lotto>,
+            winningNumbers: WinningNumbers,
+        ): Map<Int, Int> {
```
다른 개발자들도 `Map<Int, Int>` 가 어떤 타입인지 이해할 수 있을까요?

wisemuji replied:
다음 단계에 제시된 아래 요구 사항을 반영하면서 개선될 수 있을 것 같아요 🙂 

> Enum 클래스를 적용해 프로그래밍을 구현한다.

wisemuji commented on src/test/kotlin/lotto/BankTest.kt:
```diff
@@ -0,0 +1,48 @@
+import org.assertj.core.api.Assertions.entry
+import org.junit.jupiter.api.Test
+
+class BankTest {
```
테스트에서 전제 조건과, 액션, 의도된 행동(assertion) 등을 더 가시적으로 드러낼 수 있는 방법은 무엇일까요?

여러 가지 방법이 있겠지만 Given-When-Then  형식이 대표적입니다.
- https://martinfowler.com/bliki/GivenWhenThen.html

Kotest에서는 [Behavior Spec](https://kotest.io/docs/framework/testing-styles.html#behavior-spec)으로 이를 강제하기도 합니다.
꼭 반영하지 않으시고 참고만 해주셔도 좋습니다.

wisemuji reviewed (approved):
안녕하세요 희상님! 리뷰 반영을 잘 해주셨습니다.

이전에 비해서 객체 간의 역할과 책임이 좀 더 분명하게 나뉘어졌어요. 👍 
몇 가지 고민해볼 내용을 남겼는데요, 다음 단계 미션 진행해주시면서 같이 반영해주시면 됩니다.

다음 미션도 화이팅입니다!
